### ISSUE TITLE
Compiler doesn't correctly report error for applying `(?)` to non-generic types

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#23346]
In the following code, type `R` is not generic, yet the compiler doesn't complain about the application of `(?)` to it.  I believe it should.

```chapel
record R {
}

proc foo(x: R(?)) {
  writeln("In foo");
}

var r: R;
foo(r);
```

Note that this behavior predates @mppf's recent work on generic formal types in https://github.com/chapel-lang/chapel/pull/23291 and that I just happened to bump into it tonight.

**Comment by bradcray** [comment]
@mppf: I took a quick look at `main` to see whether I could easily determine where a check might be placed for this based on the changes in #23291, but enough had changed since I'd worked on it that it wasn't as obvious to me as I was hoping.  Is it to you?  (specifically, I was imagining that in places where we were checking for the absence of `(?)`-style information, we could flip it and check for the presence on a non-generic type.

(That said, I also remain a little unclear about how we're doing the warnings in normalize.cpp since I thought we couldn't generally know whether a type was generic until resolution).

**Comment by mppf** [comment]
@bradcray - I've fixed this already in #23356.

> (That said, I also remain a little unclear about how we're doing the warnings in normalize.cpp since I thought we couldn't generally know whether a type was generic until resolution).

Yes, `hack_resolve_types` is a hack.

But more specifically for this case, we have some code in normalize to mark variables with `FLAG_MARKED_GENERIC` and to remove the CallExpr for `(?)` in some cases. So we needed to add checking that, if we have `FLAG_MARKED_GENERIC`, the formal type is not concrete.
